content/blog/2016/2016-07-15-pipline-notifications.adoc

```diff
@@ -0,0 +1,306 @@
+Pipeline with an HTML publisher for code coverage. This week, I'd like to make
+Jenkins to notify me when builds start and when they succeed or fail.
+
+=== Setup and Configuration
+
+First, I select targets for my notifications. For this blog post, I'll use sample
+targets that I control.  I've created Slack and HipChat organizations called
+"bitwiseman", each with one member - me.  And for email I'm running a Ruby SMTP server called
+link:https://mailcatcher.me/[mailcatcher], that is perfect for local testing
+such as this.  Aside for these concessions, configuration would be much the
+same in a non-demo situation.
+
+Next, I install and add server-wide configuration for the
+link:https://wiki.jenkins-ci.org/display/JENKINS/Slack+Plugin[Slack],
+link:https://wiki.jenkins-ci.org/display/JENKINS/HipChat+Plugin[HipChat],
+and link:https://wiki.jenkins-ci.org/display/JENKINS/Email-ext+plugin[Email-ext]
+plugins.  Slack and HipChat use API tokens - both products have integration
+points on their side that make generate tokens which I copy into my Jenkins
+configuration. Mailcatcher SMTP runs locally. I just point Jenkins
+at it.
+
+Here's what the Jenkins configuration section for each of these looks like:
+
+image::/images/post-images/2016-07-15/slack-config.png[Slack Configuration, align="center"]
+
+image::/images/post-images/2016-07-15/hipchat-config.png[HipChat Configuration, align="center"]
+
+image::/images/post-images/2016-07-15/email-config.png[Email Configuration, align="center"]
+
+=== Original Pipeline
+
+Now I can start adding notification steps. The same as
+link:blog/2016/07/01/html-publisher-plugin/[last week], I'll use the
+link:blog/2016/05/31/pipeline-snippetizer/[Jenkins Pipeline Snippet Generator]
+to explore the step syntax for the notification plugins.
+
+Here's the base pipeline before I start making changes:
+
+[source,groovy]
+----
+stage 'Build'
+
+node {
+  // Checkout
+  checkout scm
+
+  // install required bundles
+  sh 'bundle install'
+
+  // build and run tests with coverage
+  sh 'bundle exec rake build spec'
+
+  // Archive the built artifacts
+  archive (includes: 'pkg/*.gem')
+
+  // publish html
+  // snippet generator doesn't include "target:"
+  // https://issues.jenkins-ci.org/browse/JENKINS-29711.
+  publishHTML (target: [
+      allowMissing: false,
+      alwaysLinkToLastBuild: false,
+      keepAll: true,
+      reportDir: 'coverage',
+      reportFiles: 'index.html',
+      reportName: "RCov Report"
+    ])
+}
+----
+
+
+=== Job Started Notification
+
+For the first change, I decide to add a "Job Started" notification.  The
+snippet generator and then reformatting makes this straightforward:
+
+[source,groovy]
+----
+node {
+
+  notifyStarted()
+
+  /* ... existing build steps ... */
+}
+
+def notifyStarted() {
+  // send to Slack
+  slackSend (color: '#FFFF00', message: "STARTED: Job '${env.JOB_NAME} [${env.BUILD_NUMBER}]' (${env.BUILD_URL})")
```

Comment:
I think this is the first blog post, at least that I can remember, where we show Groovy string interpolation. I think a brief mention of it, with a link to the [Groovy docs](http://docs.groovy-lang.org/latest/html/documentation/index.html#_string_interpolation), would be useful.

Something along the lines of. "Since Pipeline is a Groovy-based DSL, we can use all kinds of variables when we generate our notification messages to make them more informative/unique/speshul/etc"